server/controllers/finance/patientInvoice.js

```diff
@@ -173,30 +182,96 @@ function postSaleRecord(saleUuid, caution, userId) {
   return deferred.promise;
 }
 
-/** 
+/**
  * Utility method to ensure patient invoice lines reference sale.
- * @param {Object} saleItems - An Array of all sale items to be written 
+ * @param {Object} saleItems - An Array of all sale items to be written
  * @param {string} saleUuid - UUID of referenced patient invoice
  * @returns {Object} An Array of all sale items with guaranteed UUIDs and Patient Invoice references
  */
-function linkSaleItems(saleItems, saleUuid) { 
-  return saleItems.map(function (saleItem) { 
-     
+function linkSaleItems(saleItems, saleUuid) {
+  return saleItems.map(function (saleItem) {
+
     saleItem.uuid = saleItem.uuid || uuid();
     saleItem.sale_uuid = saleUuid;
-   
+
     // Collapse sale item into Array to be inserted into database
-    return Object.keys(saleItem).map(function (key) { 
+    return Object.keys(saleItem).map(function (key) {
       return saleItem[key];
     });
   });
 }
 
 /**
- * Utility method - determine if an array has no elements 
- * @param {Object} array - Array to be tested 
- * @returns {Boolean} Value reflecting if the array is empty or not (true or false respectively)
+ * Searches for a sale by query parameters provided.
+ *
+ * GET sales/search
  */
-function isEmpty(array) {
-  return array.length === 0;
+function search(req, res, next) {
+  'use strict';
+
+  var sql =
+    'SELECT sale.uuid, sale.project_id, CONCAT(project.abbr, sale.reference) AS reference, ' +
+      'sale.cost, sale.currency_id, sale.debitor_uuid, sale.seller_id, sale.discount, ' +
+      'sale.invoice_date, sale.note, sale.posted, sale.is_distributable ' +
+    'FROM sale JOIN project ON project.id = sale.project_id ' +
+    'WHERE ';
+
+  var conditions = [];
+
+  // look through the query string and template their key/values to the SQL query
+  var tmpl = Object.keys(req.query).map(function (key) {
```

Comment:
This method has enough responsibilities to consider splitting it into one or two functions. If this pattern is used across a number of services (which I believe it is) it could be moved to a library to help make this code more readable.

This is an iterative process so will not block this pull request.